[DataPipe] Improving debug message when exceptions are raised within IterDataPipe

NivekT opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #75618

This PR aims to improve the error message coming from `IterDataPipe`. Fixes the "differentiate DataPipe instances" part of https://github.com/pytorch/data/issues/284.

For example, for this code snippet:
```python
from torchdata.datapipes.iter import IterableWrapper, Mapper

def no_op(x):
    return x

def exception_when_one(x):
    if x == 1:
        raise RuntimeError("x cannot equal to 1.")
    return x

map_dp = IterableWrapper(range(10))
map_dp = Mapper(map_dp, no_op)
map_dp = Mapper(map_dp, exception_when_one)
map_dp = Mapper(map_dp, no_op)
print(list(map_dp))
```

Here is the trace prior to this PR. It doesn't clearly state that a DataPipe has encountered an error and which DataPipe is responsible.
```
Traceback (most recent call last):
  File "/Users/ktse/scratch/debugability.py", line 63, in <module>
    print(list(map_dp))
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/_typing.py", line 369, in wrap_generator
    response = gen.send(request)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/callable.py", line 110, in __iter__
    for data in self.datapipe:
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/_typing.py", line 369, in wrap_generator
    response = gen.send(request)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/callable.py", line 111, in __iter__
    yield self._apply_fn(data)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/callable.py", line 76, in _apply_fn
    return self.fn(data)
  File "/Users/ktse/scratch/debugability.py", line 55, in exception_when_one
    raise RuntimeError("x cannot equal to 1.")
RuntimeError: x cannot equal to 1.
```

Here is the trace with this PR. In this version, it more clearly shows that a `MapperIterDataPipe` is throwing the exception and what the input arguments to the DataPipe are:
```
Traceback (most recent call last):
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/_typing.py", line 395, in wrap_generator
    response = gen.send(request)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/callable.py", line 111, in __iter__
    yield self._apply_fn(data)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/callable.py", line 76, in _apply_fn
    return self.fn(data)
  File "/Users/ktse/scratch/debugability.py", line 55, in exception_when_one
    raise RuntimeError("x cannot equal to 1.")
RuntimeError: x cannot equal to 1.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/ktse/scratch/debugability.py", line 63, in <module>
    print(list(map_dp))
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/_typing.py", line 395, in wrap_generator
    response = gen.send(request)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/callable.py", line 110, in __iter__
    for data in self.datapipe:
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/_typing.py", line 406, in wrap_generator
    raise DataPipeException(msg, e) from e
torch.utils.data.datapipes._typing.DataPipeException: thrown by __iter__ of MapperIterDataPipe(datapipe=MapperIterDataPipe, fn=exception_when_one, input_col=NoneType, output_col=NoneType)
```

Improvements to make: make the traceback shorter, such as by skipping over parts related to `response = gen.send(request)` or `for data in self.datapipe`.

NivekT committed:
[DataPipe] Improving debug message when exceptions are raised within IterDataPipe

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Improving debug message when exceptions are raised within IterDataPipe"




[ghstack-poisoned]

NivekT commented on torch/utils/data/datapipes/_typing.py:
```diff
@@ -369,6 +395,15 @@ def wrap_generator(*args, **kwargs):
+                # Short version, but requires custom __repr__ for each DataPipe class
+                # msg = f"thrown by {datapipe}"
+                # Long version with all input arguments
+                msg = f"thrown by __iter__ of {datapipe.__class__.__name__}({get_input_args(datapipe)})"
```
Thoughts on listing out all the input arguments? I think it makes it obvious to the users which DataPipe is responsible at the cost of having a longer message.

NivekT commented on torch/utils/data/datapipes/_typing.py:
```diff
@@ -337,6 +337,32 @@ def _hash_(self):
+    return ', '.join([f'{name}={value}' for name, value in result])
+
+
+class DataPipeException(Exception):
```
I am adding a custom Exception here so that when the exception propagates through the DataPipes, it will not need to be re-raised and can pass through.

Perhaps it can be helpful in cutting down the traceback as well? Any idea?

NivekT replied:
Seems like starting with Python 3.7, we can use the `with_traceback` method. I will play around with that

NivekT committed:
Update on "[DataPipe] Improving debug message when exceptions are raised within IterDataPipe"


This PR aims to improve the error message coming from `IterDataPipe`.

For example, for this code snippet:
```python
from torchdata.datapipes.iter import IterableWrapper, Mapper

def no_op(x):
    return x

def exception_when_one(x):
    if x == 1:
        raise RuntimeError("x cannot equal to 1.")
    return x

map_dp = IterableWrapper(range(10))
map_dp = Mapper(map_dp, no_op)
map_dp = Mapper(map_dp, exception_when_one)
map_dp = Mapper(map_dp, no_op)
print(list(map_dp))
```

Here is the trace prior to this PR. It doesn't clearly state that a DataPipe has encountered an error and which DataPipe is responsible.
```
Traceback (most recent call last):
  File "/Users/ktse/scratch/debugability.py", line 63, in <module>
    print(list(map_dp))
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/_typing.py", line 369, in wrap_generator
    response = gen.send(request)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/callable.py", line 110, in __iter__
    for data in self.datapipe:
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/_typing.py", line 369, in wrap_generator
    response = gen.send(request)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/callable.py", line 111, in __iter__
    yield self._apply_fn(data)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/callable.py", line 76, in _apply_fn
    return self.fn(data)
  File "/Users/ktse/scratch/debugability.py", line 55, in exception_when_one
    raise RuntimeError("x cannot equal to 1.")
RuntimeError: x cannot equal to 1.
```

Here is the trace with this PR. In this version, it more clearly shows that a `MapperIterDataPipe` is throwing the exception and what the input arguments to the DataPipe are:
```
Traceback (most recent call last):
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/_typing.py", line 395, in wrap_generator
    response = gen.send(request)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/callable.py", line 111, in __iter__
    yield self._apply_fn(data)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/callable.py", line 76, in _apply_fn
    return self.fn(data)
  File "/Users/ktse/scratch/debugability.py", line 55, in exception_when_one
    raise RuntimeError("x cannot equal to 1.")
RuntimeError: x cannot equal to 1.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/ktse/scratch/debugability.py", line 63, in <module>
    print(list(map_dp))
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/_typing.py", line 395, in wrap_generator
    response = gen.send(request)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/callable.py", line 110, in __iter__
    for data in self.datapipe:
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/_typing.py", line 406, in wrap_generator
    raise DataPipeException(msg, e) from e
torch.utils.data.datapipes._typing.DataPipeException: thrown by __iter__ of MapperIterDataPipe(datapipe=MapperIterDataPipe, fn=exception_when_one, input_col=NoneType, output_col=NoneType)
```

Improvements to make: make the traceback shorter, such as by skipping over parts related to `response = gen.send(request)` or `for data in self.datapipe`.

[ghstack-poisoned]

ejguan commented:
I have one concern on the new traceback. The original traceback is all the way back to `print(list(map_dp))`. However, the new traceback is somehow split the traceback into two parts. I find it hard to understand which part it the actual root cause at the begining.

Can we let `DataPipeException` to inherit the message and traceback from original one and appending the message `msg = f"thrown by __iter__ of {datapipe.__class__.__name__}({_generate_input_args_string(datapipe)})"`

NivekT commented:
I just updated the code. This will now be the error message:
```
Traceback (most recent call last):
  File "/Users/ktse/scratch/debugability.py", line 63, in <module>
    print(list(map_dp))
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/_typing.py", line 396, in wrap_generator
    response = gen.send(request)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/callable.py", line 110, in __iter__
    for data in self.datapipe:
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/_typing.py", line 396, in wrap_generator
    response = gen.send(request)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/callable.py", line 111, in __iter__
    yield self._apply_fn(data)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/callable.py", line 76, in _apply_fn
    return self.fn(data)
  File "/Users/ktse/scratch/debugability.py", line 55, in exception_when_one
    raise RuntimeError("x cannot equal to 1.")
RuntimeError: x cannot equal to 1.
 This exception is thrown by __iter__ of MapperIterDataPipe(datapipe=MapperIterDataPipe, fn=exception_when_one, input_col=NoneType, output_col=NoneType)
```

Basically adding an extra line but not anything else. No custom exception is needed either.

NivekT committed:
Update on "[DataPipe] Improving debug message when exceptions are raised within IterDataPipe"


This PR aims to improve the error message coming from `IterDataPipe`. Fixes the "differentiate DataPipe instances" part of https://github.com/pytorch/data/issues/284.

For example, for this code snippet:
```python
from torchdata.datapipes.iter import IterableWrapper, Mapper

def no_op(x):
    return x

def exception_when_one(x):
    if x == 1:
        raise RuntimeError("x cannot equal to 1.")
    return x

map_dp = IterableWrapper(range(10))
map_dp = Mapper(map_dp, no_op)
map_dp = Mapper(map_dp, exception_when_one)
map_dp = Mapper(map_dp, no_op)
print(list(map_dp))
```

Here is the trace prior to this PR. It doesn't clearly state that a DataPipe has encountered an error and which DataPipe is responsible.
```
Traceback (most recent call last):
  File "/Users/ktse/scratch/debugability.py", line 63, in <module>
    print(list(map_dp))
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/_typing.py", line 369, in wrap_generator
    response = gen.send(request)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/callable.py", line 110, in __iter__
    for data in self.datapipe:
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/_typing.py", line 369, in wrap_generator
    response = gen.send(request)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/callable.py", line 111, in __iter__
    yield self._apply_fn(data)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/callable.py", line 76, in _apply_fn
    return self.fn(data)
  File "/Users/ktse/scratch/debugability.py", line 55, in exception_when_one
    raise RuntimeError("x cannot equal to 1.")
RuntimeError: x cannot equal to 1.
```

Here is the trace with this PR. In this version, it more clearly shows that a `MapperIterDataPipe` is throwing the exception and what the input arguments to the DataPipe are:
```
Traceback (most recent call last):
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/_typing.py", line 395, in wrap_generator
    response = gen.send(request)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/callable.py", line 111, in __iter__
    yield self._apply_fn(data)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/callable.py", line 76, in _apply_fn
    return self.fn(data)
  File "/Users/ktse/scratch/debugability.py", line 55, in exception_when_one
    raise RuntimeError("x cannot equal to 1.")
RuntimeError: x cannot equal to 1.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/ktse/scratch/debugability.py", line 63, in <module>
    print(list(map_dp))
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/_typing.py", line 395, in wrap_generator
    response = gen.send(request)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/callable.py", line 110, in __iter__
    for data in self.datapipe:
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/_typing.py", line 406, in wrap_generator
    raise DataPipeException(msg, e) from e
torch.utils.data.datapipes._typing.DataPipeException: thrown by __iter__ of MapperIterDataPipe(datapipe=MapperIterDataPipe, fn=exception_when_one, input_col=NoneType, output_col=NoneType)
```

Improvements to make: make the traceback shorter, such as by skipping over parts related to `response = gen.send(request)` or `for data in self.datapipe`.

[ghstack-poisoned]

ejguan commented on torch/utils/data/datapipes/_typing.py:
```diff
@@ -375,6 +402,15 @@ def wrap_generator(*args, **kwargs):
+                datapipe = args[0]
+                msg = "thrown by __iter__ of"
+                full_msg = f"{msg} {datapipe.__class__.__name__}({_generate_input_args_string(datapipe)})"
+                if len(e.args) >= 1 and msg not in e.args[0]:
+                    e.args = (e.args[0] + f'\n This exception is {full_msg}',) + e.args[1:]
```
I have one concern on this approach. Seeing from python document (https://docs.python.org/3/library/exceptions.html#BaseException.args), does this work for `OSError`?

ejguan replied:
It seems working. NVM.

ejguan commented on torch/utils/data/datapipes/_typing.py:
```diff
@@ -375,6 +402,15 @@ def wrap_generator(*args, **kwargs):
+                msg = "thrown by __iter__ of"
+                full_msg = f"{msg} {datapipe.__class__.__name__}({_generate_input_args_string(datapipe)})"
+                if len(e.args) >= 1 and msg not in e.args[0]:
+                    e.args = (e.args[0] + f'\n This exception is {full_msg}',) + e.args[1:]
```
Nit for the message
```suggestion
                    e.args = (e.args[0] + f'\nThis exception is {full_msg}',) + e.args[1:]
```

ejguan commented on torch/utils/data/datapipes/_typing.py:
```diff
@@ -337,6 +337,33 @@ def _hash_(self):
+    """
+    if inspect.isfunction(obj):
+        return obj.__name__
+    try:
+        return str(obj.__class__.__qualname__)
+    except Exception as _:
+        return str(obj)
```
Wondering why we want to return the class of object rather than object itself?
Like `None` value, the Error message is `NoneType` currently. IMHO, `None` would be more better.

NivekT replied:
It typically gives us a shorter and cleaner name (especially for functions/classes). The `NoneType` behavior may be one of the few cases where it gives a longer name

NivekT replied:
Let me have a look at the behavior of a few more types and see what may be best

ejguan replied:
For example, providing the type for `integer`, `float` and other common data stuctures may causes users more confusion on which `DataPipe` caused such issue.

NivekT replied:
I updated this section of the code to only show the simplified class name when the default string is something like `<very.long.path.to.ClassName object at 0x000000000000>`.

ejguan replied:
Please correct me if I am wrong. Based on the comment in the function, does`<very.long.path.to.ClassName object at 0x000000000000>` only refer to `DataPipe` or refer to any other object?
If it's former, why not we add a `__str__` method to `DataPipe` class to always only return the DataPipe's name?

NivekT replied:
It is being applied to any object. My thinking is that `<...>` isn't very informative and just the class name provide the same amount of information.

We definitely can only apply that logic to `DataPipe` and have that as a `__str__` method.

NivekT replied:
I updated the PR so that simplified object string only affects DataPipes when exceptions are raised within DataPipes.

ejguan reviewed (commented):
Really appreciate digging into this. The structure of the Error message is good. But, I still have a few comments below.

NivekT committed:
Update on "[DataPipe] Improving debug message when exceptions are raised within IterDataPipe"


This PR aims to improve the error message coming from `IterDataPipe`. Fixes the "differentiate DataPipe instances" part of https://github.com/pytorch/data/issues/284.

For example, for this code snippet:
```python
from torchdata.datapipes.iter import IterableWrapper, Mapper

def no_op(x):
    return x

def exception_when_one(x):
    if x == 1:
        raise RuntimeError("x cannot equal to 1.")
    return x

map_dp = IterableWrapper(range(10))
map_dp = Mapper(map_dp, no_op)
map_dp = Mapper(map_dp, exception_when_one)
map_dp = Mapper(map_dp, no_op)
print(list(map_dp))
```

Here is the trace prior to this PR. It doesn't clearly state that a DataPipe has encountered an error and which DataPipe is responsible.
```
Traceback (most recent call last):
  File "/Users/ktse/scratch/debugability.py", line 63, in <module>
    print(list(map_dp))
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/_typing.py", line 369, in wrap_generator
    response = gen.send(request)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/callable.py", line 110, in __iter__
    for data in self.datapipe:
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/_typing.py", line 369, in wrap_generator
    response = gen.send(request)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/callable.py", line 111, in __iter__
    yield self._apply_fn(data)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/callable.py", line 76, in _apply_fn
    return self.fn(data)
  File "/Users/ktse/scratch/debugability.py", line 55, in exception_when_one
    raise RuntimeError("x cannot equal to 1.")
RuntimeError: x cannot equal to 1.
```

Here is the trace with this PR. In this version, it more clearly shows that a `MapperIterDataPipe` is throwing the exception and what the input arguments to the DataPipe are:
```
Traceback (most recent call last):
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/_typing.py", line 395, in wrap_generator
    response = gen.send(request)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/callable.py", line 111, in __iter__
    yield self._apply_fn(data)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/callable.py", line 76, in _apply_fn
    return self.fn(data)
  File "/Users/ktse/scratch/debugability.py", line 55, in exception_when_one
    raise RuntimeError("x cannot equal to 1.")
RuntimeError: x cannot equal to 1.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/ktse/scratch/debugability.py", line 63, in <module>
    print(list(map_dp))
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/_typing.py", line 395, in wrap_generator
    response = gen.send(request)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/callable.py", line 110, in __iter__
    for data in self.datapipe:
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/_typing.py", line 406, in wrap_generator
    raise DataPipeException(msg, e) from e
torch.utils.data.datapipes._typing.DataPipeException: thrown by __iter__ of MapperIterDataPipe(datapipe=MapperIterDataPipe, fn=exception_when_one, input_col=NoneType, output_col=NoneType)
```

Improvements to make: make the traceback shorter, such as by skipping over parts related to `response = gen.send(request)` or `for data in self.datapipe`.

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Improving debug message when exceptions are raised within IterDataPipe"


This PR aims to improve the error message coming from `IterDataPipe`. Fixes the "differentiate DataPipe instances" part of https://github.com/pytorch/data/issues/284.

For example, for this code snippet:
```python
from torchdata.datapipes.iter import IterableWrapper, Mapper

def no_op(x):
    return x

def exception_when_one(x):
    if x == 1:
        raise RuntimeError("x cannot equal to 1.")
    return x

map_dp = IterableWrapper(range(10))
map_dp = Mapper(map_dp, no_op)
map_dp = Mapper(map_dp, exception_when_one)
map_dp = Mapper(map_dp, no_op)
print(list(map_dp))
```

Here is the trace prior to this PR. It doesn't clearly state that a DataPipe has encountered an error and which DataPipe is responsible.
```
Traceback (most recent call last):
  File "/Users/ktse/scratch/debugability.py", line 63, in <module>
    print(list(map_dp))
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/_typing.py", line 369, in wrap_generator
    response = gen.send(request)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/callable.py", line 110, in __iter__
    for data in self.datapipe:
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/_typing.py", line 369, in wrap_generator
    response = gen.send(request)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/callable.py", line 111, in __iter__
    yield self._apply_fn(data)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/callable.py", line 76, in _apply_fn
    return self.fn(data)
  File "/Users/ktse/scratch/debugability.py", line 55, in exception_when_one
    raise RuntimeError("x cannot equal to 1.")
RuntimeError: x cannot equal to 1.
```

Here is the trace with this PR. In this version, it more clearly shows that a `MapperIterDataPipe` is throwing the exception and what the input arguments to the DataPipe are:
```
Traceback (most recent call last):
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/_typing.py", line 395, in wrap_generator
    response = gen.send(request)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/callable.py", line 111, in __iter__
    yield self._apply_fn(data)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/callable.py", line 76, in _apply_fn
    return self.fn(data)
  File "/Users/ktse/scratch/debugability.py", line 55, in exception_when_one
    raise RuntimeError("x cannot equal to 1.")
RuntimeError: x cannot equal to 1.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/ktse/scratch/debugability.py", line 63, in <module>
    print(list(map_dp))
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/_typing.py", line 395, in wrap_generator
    response = gen.send(request)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/callable.py", line 110, in __iter__
    for data in self.datapipe:
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/_typing.py", line 406, in wrap_generator
    raise DataPipeException(msg, e) from e
torch.utils.data.datapipes._typing.DataPipeException: thrown by __iter__ of MapperIterDataPipe(datapipe=MapperIterDataPipe, fn=exception_when_one, input_col=NoneType, output_col=NoneType)
```

Improvements to make: make the traceback shorter, such as by skipping over parts related to `response = gen.send(request)` or `for data in self.datapipe`.

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Improving debug message when exceptions are raised within IterDataPipe"


This PR aims to improve the error message coming from `IterDataPipe`. Fixes the "differentiate DataPipe instances" part of https://github.com/pytorch/data/issues/284.

For example, for this code snippet:
```python
from torchdata.datapipes.iter import IterableWrapper, Mapper

def no_op(x):
    return x

def exception_when_one(x):
    if x == 1:
        raise RuntimeError("x cannot equal to 1.")
    return x

map_dp = IterableWrapper(range(10))
map_dp = Mapper(map_dp, no_op)
map_dp = Mapper(map_dp, exception_when_one)
map_dp = Mapper(map_dp, no_op)
print(list(map_dp))
```

Here is the trace prior to this PR. It doesn't clearly state that a DataPipe has encountered an error and which DataPipe is responsible.
```
Traceback (most recent call last):
  File "/Users/ktse/scratch/debugability.py", line 63, in <module>
    print(list(map_dp))
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/_typing.py", line 369, in wrap_generator
    response = gen.send(request)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/callable.py", line 110, in __iter__
    for data in self.datapipe:
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/_typing.py", line 369, in wrap_generator
    response = gen.send(request)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/callable.py", line 111, in __iter__
    yield self._apply_fn(data)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/callable.py", line 76, in _apply_fn
    return self.fn(data)
  File "/Users/ktse/scratch/debugability.py", line 55, in exception_when_one
    raise RuntimeError("x cannot equal to 1.")
RuntimeError: x cannot equal to 1.
```

Here is the trace with this PR. In this version, it more clearly shows that a `MapperIterDataPipe` is throwing the exception and what the input arguments to the DataPipe are:
```
Traceback (most recent call last):
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/_typing.py", line 395, in wrap_generator
    response = gen.send(request)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/callable.py", line 111, in __iter__
    yield self._apply_fn(data)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/callable.py", line 76, in _apply_fn
    return self.fn(data)
  File "/Users/ktse/scratch/debugability.py", line 55, in exception_when_one
    raise RuntimeError("x cannot equal to 1.")
RuntimeError: x cannot equal to 1.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/ktse/scratch/debugability.py", line 63, in <module>
    print(list(map_dp))
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/_typing.py", line 395, in wrap_generator
    response = gen.send(request)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/callable.py", line 110, in __iter__
    for data in self.datapipe:
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/_typing.py", line 406, in wrap_generator
    raise DataPipeException(msg, e) from e
torch.utils.data.datapipes._typing.DataPipeException: thrown by __iter__ of MapperIterDataPipe(datapipe=MapperIterDataPipe, fn=exception_when_one, input_col=NoneType, output_col=NoneType)
```

Improvements to make: make the traceback shorter, such as by skipping over parts related to `response = gen.send(request)` or `for data in self.datapipe`.

[ghstack-poisoned]

ejguan reviewed (approved):
Thanks for adding this patch. LGTM

ejguan commented on torch/utils/data/datapipes/_typing.py:
```diff
@@ -375,6 +407,15 @@ def wrap_generator(*args, **kwargs):
+                # TODO: Simplify the traceback message to skip over `response = gen.send(None)`
+                #       Part of https://github.com/pytorch/data/issues/284
+                datapipe = args[0]
+                msg = "thrown by __iter__ of"
+                full_msg = f"{msg} {datapipe.__class__.__name__}({_generate_input_args_string(datapipe)})"
```
When the Error comes from singleton iterator. This part of message becomes duplicated.
You can run the following example:
```py
source_dp = IterableWrapper(range(10))
it1 = iter(source_dp)
next(it1)
it2 = iter(source_dp)
next(it2)
next(it1)
```
The Error message is:
```py
RuntimeError: This iterator has been invalidated because another iterator has been created from the same IterDataPipe: IterableWrapperIterDataPipe(deepcopy=True, iterable=range(0, 10))
This may be caused multiple references to the same IterDataPipe. We recommend using `.fork()` if that is necessary.
For feedback regarding this single iterator per IterDataPipe constraint, feel free to comment on this issue: https://github.com/pytorch/data/issues/45.
This exception is thrown by __iter__ of IterableWrapperIterDataPipe(deepcopy=True, iterable=range(0, 10))
```

cc: @NivekT

NivekT replied:
You are saying that we don't want `IterableWrapperIterDataPipe(deepcopy=True, iterable=range(0, 10))` to appear twice?

NivekT replied:
I am open to that idea, but it looks fine to me.

ejguan replied:
That would be the ideal case, IMO. We use `This exception is thrown by __iter__ of IterableWrapperIterDataPipe(deepcopy=True, iterable=range(0, 10))` to indicate the Error comes from internal logic of `__iter__`. But, the actual Error comes from the duplicate iterators. I think this message is a little misleading.

NivekT replied:
I see your point, I can fix that.

ejguan replied:
Thank you. Not high-pri though.
